File: leetcode/solutions/0543.DiameterOfBinaryTree/diameterOfBinaryTree_test.ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";
import diameterOfBinaryTree from "./diameterOfBinaryTree.ts";
import {
  createBinaryTreeNode,
  BinaryTreeNode,
} from "../../../datastructure/BinaryTreeNode.ts";

Deno.test("0543. Diameter of Binary Tree", () => {
  type test = [BinaryTreeNode<number> | null, number];
  const tests: test[] = [
    [null, 0],
    [createBinaryTreeNode([1]), 0],
    [createBinaryTreeNode([1, 2]), 1],
    [createBinaryTreeNode([1, 2, 3]), 2],
    [createBinaryTreeNode([1, null, 2]), 1],
    [createBinaryTreeNode([1, 2, null, 3]), 2],
  ];

  for (const [arg1, want] of tests) {
    assertEquals(diameterOfBinaryTree(arg1), want);
  }
});

File: leetcode/solutions/0543.DiameterOfBinaryTree/diameterOfBinaryTree_test.go
package solution

import (
	"testing"

	ds "github.com/yukisato/leetcode-util/datastructure"
)

func Test_0543_DiameterOfBinaryTree(t *testing.T) {
	t.Parallel()

	tests := []struct {
		in  *TreeNode
		out int
	}{
		{nil, 0},
		{ds.CreateTreeNodeFromString([]string{"1"}), 0},
		{ds.CreateTreeNodeFromString([]string{"1", "null", "2"}), 1},
		{ds.CreateTreeNodeFromString([]string{"1", "2", "3", "4", "5"}), 3},
		{ds.CreateTreeNodeFromString([]string{"2", "3", "null", "1"}), 2},
	}

	for _, tt := range tests {
		tt := tt

		t.Run("Test", func(t *testing.T) {
			t.Parallel()

			got := diameterOfBinaryTree(tt.in)

			if got != tt.out {
				t.Errorf("got %#v want %#v", got, tt.out)
			}
		})
	}
}
